### PR TITLE
Azure Event Hub receiver Semantic Convention translator

### PR DESCRIPTION
**Pull request by ThatRendle**
**Description:**

This PR changes the translator used by the Azure Event Hubs receiver to change property names on log records and metrics to the equivalent OTel Semantic Conventions names, making it easier to query logs and metrics ingested in this way alongside other logs and metrics. 

**Testing:** WIP.

**Documentation:** WIP

### CODE REVIEW

File: pkg/translator/azure/property_names.go
Comment (TylerHelmuth) [inline pkg/translator/azure/property_names.go:3]: This can be made a const

File: pkg/translator/azure/property_names.go
Comment (TylerHelmuth) [inline pkg/translator/azure/property_names.go:39]: Let's rename this `resourceLogKeyToSemConvKey` so it isn't exported.

File: pkg/translator/azure/property_names.go
Comment (TylerHelmuth) [inline pkg/translator/azure/property_names.go:40]: We can use the `if mapping, ok := mappings[category]; ok {` syntax here and below instead. Personally I prefer

```go
mapping, ok := mappings[category]
if ok {
...
}
```

File: pkg/translator/azure/resourcelogs_to_logs.go
Comment (TylerHelmuth) [inline pkg/translator/azure/resourcelogs_to_logs.go:221]: What does `azure.properties` represent? Properties that don't have a semantic convention mapping?

File: pkg/translator/azure/resourcelogs_to_logs.go
Comment (TylerHelmuth) [inline pkg/translator/azure/resourcelogs_to_logs.go:219]: Instead of doing this inside the loop we could instantiate the map before the loop, inside the else block add properties, and then when the loop is finished check if the length of the map is greater than 0 and add it to `azure.properties` if true.

File: pkg/translator/azure/property_names.go
Comment (TylerHelmuth) [reply pkg/translator/azure/property_names.go:3]: I am wrong about this

File: pkg/translator/azure/resourcelogs_to_logs.go
Comment (markrendle) [reply pkg/translator/azure/resourcelogs_to_logs.go:219]: I was trying to avoid the allocation unless needed; is that an over-optimization?

Comment (markrendle) [reply pkg/translator/azure/resourcelogs_to_logs.go:219]: Ah. An empty map won't be allocated on the heap, right?

File: pkg/translator/azure/resourcelogs_to_logs.go
Comment (TylerHelmuth) [reply pkg/translator/azure/resourcelogs_to_logs.go]: Sweet

File: pkg/translator/azure/resourcelogs_to_logs.go
Comment (TylerHelmuth) [inline pkg/translator/azure/resourcelogs_to_logs.go:223]: we can use the `azureProperties` const for this key

**Review by TylerHelmuth — COMMENTED** [review]
Please add some new test scenarios

**Comment by cparkins** [comment]
I like this idea and I think it makes a lot of sense to try to map to Semantic Conventions were possible. Can you make sure an Issue is created to track this against?

File: pkg/translator/azure/property_names.go
Comment (lmolkova) [inline pkg/translator/azure/property_names.go:8]: I believe [network.protocol.version](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/attributes-registry/network.md) would be the right match

File: pkg/translator/azure/property_names.go
Comment (lmolkova) [inline pkg/translator/azure/property_names.go:6]: nit: you might consider mapping it to `az.service_request_id` - we use it in the Azure SDKs to record unique id returned by the service in response - https://github.com/Azure/azure-sdk/blob/main/docs/tracing/distributed-tracing-conventions.md#http-client-spans

File: pkg/translator/azure/property_names.go
Comment (lmolkova) [inline pkg/translator/azure/property_names.go:46]: no strong opinion, but it seems `http.server.request.duration` was used above, seems like  this one matches the `http.client.request.duration` (which is in seconds and is a metric, so if we map, I think we should also convert to double seconds)

File: pkg/translator/azure/property_names.go
Comment (lmolkova) [inline pkg/translator/azure/property_names.go:48]: in theory we can reuse [`dns.lookup.duration`](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/dns/dns-metrics.md), but need to convert to seconds

File: pkg/translator/azure/property_names.go
Comment (markrendle) [reply pkg/translator/azure/property_names.go]: Yes, what you get for copying and pasting 😊

**Comment by atoulme** [comment]
Can you please also contribute those transformations to https://github.com/open-telemetry/semantic-conventions so we have it in the spec?

**Review by MikeGoldsmith — APPROVED** [review]
This looks great and will greatly improve consistency for Azure events - thanks @markrendle 🚀

**Comment by crobert-1** [comment]
Going along with @atoulme's request, I think it may be good to block this PR until the suggested conventions here have been approved and merged into the semantic conventions repo, as any changes requested there will need to be reflected in this package (and PR). The problem with going forward with this PR before semantic conventions are finalized, would be the decisions there could end up being breaking changes here (if this gets merged and released).

Let me know if I've misunderstood the nature of this change though 👍

**Comment by TylerHelmuth** [comment]
@atoulme @crobert-1 @cparkins I will get a semconv issue/PR opened to add the net-new attribute names.  As for the mapping of azure fields to existing fields, I do not believe the semantic convention currently defines those.  [The closest I know about is are these appendices ](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/data-model-appendix.md#google-cloud-logging).  I will open a issue/PR for that as well.

File: pkg/translator/azure/resourcelogs_to_logs.go
Comment (TylerHelmuth) [inline pkg/translator/azure/resourcelogs_to_logs.go:197]: Thinking about these changes some more, they end up being a breaking change regardless of whether or not the semconv have approved the attribute names.  This change moves the data out of the `azure.properties` and into new, semantically relevant, fields.  The result is the user needs to good look for their data in a new place - a better place, but new nevertheless.

Here is my proposal for how to move these breaking changes forward to give users time to adjust and and time for the semconv to discuss the names. Lets add 2 new feature gates that will control how the data is produced.
1. A feature gate that, when enabled, will translate the data in the old way and the new way.
2. A feature gate that, when enabled, will translate the data in only the new way.

If both feature gates are enabled, the first gate will take precedence.

Comment (martinjt) [reply pkg/translator/azure/resourcelogs_to_logs.go:197]: I'd suggest that this is a new translator, so it's opt-in in the config for this work to come into effect.

Comment (TylerHelmuth) [reply pkg/translator/azure/resourcelogs_to_logs.go:197]: A new translator (or a new flow in this azure translator) works too.  The eventhub receiver will need 2 new formats to help with the transition tho, 1 format that emits only the new data, and 1 format that emits the new data and the old data.

Comment (TylerHelmuth) [reply pkg/translator/azure/resourcelogs_to_logs.go:197]: @atoulme @crobert-1 please weigh in as code owners on how you'd like to see these changes integrated.

Comment (TylerHelmuth) [reply pkg/translator/azure/resourcelogs_to_logs.go:197]: Also @djaglowski

Comment (cparkins) [reply pkg/translator/azure/resourcelogs_to_logs.go]: @TylerHelmuth I believe that providing a way to only emit old or new is sufficient. I can see an argument for both; but I do agree this could be considered a breaking change.

**Comment by MikeGoldsmith** [comment]
This is still a WIP. The work to add semmantic conventions for the new azure namespace is on going.

**Comment by MovieStoreGuy** [comment]
Is it worth converting this PR to draft until it is ready for review?

**Comment by MikeGoldsmith** [comment]
Yep, I think that's wise given it's blocked on semantic conventions PR:
- https://github.com/open-telemetry/semantic-conventions/pull/1028

**Comment by MikeGoldsmith** [comment]
Note to add opt-in for identity field as it contains PII data.
- https://github.com/open-telemetry/semantic-conventions/pull/1028/files#r1613532265

**Comment by crobert-1** [comment]
There isn't a good label for that, but I can add `never stale` for now.

**Comment by MikeGoldsmith** [comment]
As an update, the semconv PR has picked up again and we'll be moving to an event model instead of a generic log. Named events have the benefit of a clear schema of field names where logs do not. Hopefully that PR will wrap up soon and we can update this work to reflect the changes.

**Comment by MikeGoldsmith** [comment]
This PR can be closed now #34830 has merged.

**Comment by codeboten** [comment]
> This PR can be closed now #34830 has merged.

👍🏻

**Pull request closed without merging**
